src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,316 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
+ * to verify that client can use the certificate based on the extension policy
+ * defined by modules.
+ */
+static krb5_error_code
+authorize_cert(krb5_context context, certauth_handle *certauth_modules,
+               pkinit_kdc_context plgctx, pkinit_kdc_req_context reqctx,
+               krb5_kdcpreauth_callbacks cb, krb5_kdcpreauth_rock rock,
+               krb5_principal client)
+{
+    krb5_error_code ret;
+    certauth_handle hd;
+    struct certauth_req_opts opts;
+    krb5_boolean accepted = FALSE;
+    krb5_data cert;
+    size_t i;
+    void *db_ent = NULL;
+    char **ais = NULL, **ai = NULL;
+
+    /* Re-encode the received certificate into DER, which is extra work, but
+     * avoids creating a crypto dependency on the interface. */
+    ret = crypto_encode_der_cert(context, reqctx->cryptoctx, &cert);
+    if (ret)
+        goto cleanup;
+
+    /* Set options for the builtin module. */
+    opts.plgctx = plgctx;
+    opts.reqctx = reqctx;
+    opts.cb = cb;
+    opts.rock = rock;
+
+    db_ent = cb->client_entry(context, rock);
+
+    /*
+     * Check the certificate against each certauth module. For the certificate
+     * to be authorized at least one module must return 0 and no module returns
+     * an error code other than KRB5_PLUGIN_NO_HANDLE (pass).  Add indicators
+     * from modules that return 0 or pass.
+     */
+    ret = KRB5_PLUGIN_NO_HANDLE;
+    for (i = 0; certauth_modules != NULL && certauth_modules[i] != NULL; i++) {
+        hd = certauth_modules[i];
+        if (hd->vt.authorize == NULL)
+            continue;
+
+        ret = hd->vt.authorize(context, hd->moddata, (const uint8_t *)cert.data,
+                               cert.length, client, &opts, db_ent, &ais);
+
+        if (ret == 0)
+            accepted = TRUE;
+        else if (ret != KRB5_PLUGIN_NO_HANDLE)
+            goto cleanup;
+
+        if (ais != NULL) {
+            /* Assert authentication indicators from certificate authorization. */
+            for (ai = ais; *ai != NULL; ai++) {
+                ret = cb->add_auth_indicator(context, rock, *ai);
+                if (ret)
+                    goto cleanup;
+            }
+
+            if (hd->vt.free_ind != NULL)
+                hd->vt.free_ind(context, hd->moddata, ais);
+            else
+                free_ai_list(ais);
+            ais = NULL;
+        }
+    }
+
+    ret = accepted ? 0 : KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+
+cleanup:
+    crypto_free_cert_data(context, cert);
+    return ret;
+}
+
+static krb5_error_code
+pkinitsrv_san_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                        const uint8_t *cert, size_t cert_len,
+                        krb5_const_principal princ, const void *opts,
+                        const krb5_db_entry *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_san;
+    const struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
```

Comment:
No need to test for nullity here; it's not an optional return parameter.  (Also applies to eku module.)